README.md
# shaped_buttons

Language: [English](README.md)

Shaped Buttons is a flutter library that allows you to create parallelogram shaped featured buttons.

![](https://github.com/ibtihaajKhurram/shaped_buttons/blob/master/assets/screenshots/Screenshot_01.jpg)

- [shaped_buttons](#shapedbuttons)
  - [How to use it.](#how-to-use-it)
  - [parameters](#parameters)

##  How to use it.

the default plain parallelogram
```dart
  ParallalogramButton(
                child: Text(
                  "Regular Button",
                  style: TextStyle(color: Colors.black),
                ),
                onPressed: () {
                  print('Button Pressed');
                },
              ),
```

and you can also define custom properties.
```dart
 ParallalogramButton(
                width: 200,
                borderColor: Colors.black,
                buttonColor: Colors.yellow[700],
                borderWdth: 5,
                child: Text(
                  "Button with border",
                  style: TextStyle(color: Colors.black),
                ),
                onPressed: () {
                  print('Button Pressed');
                },
              ),
```


## parameters

| parameter                  | description                                                                           | default                                                                                                                                                                               |
| -------------------------- | ------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| child                       | child widget (usually a Text widget)                                                                   | 	This is a required parameter                                                                                                                                                                                  |
| width                       | width of button                                                                   | 120.0                                                                                                                                                                                  |
| height                       | height of button                                                                   | 50.0                                                                                                                                                                                  |
| margin                       | Margin around the button                                                                   | EdgeInsets.only(left: 1, right: 1)                                                                                                                                                                                  |
| borderWidth                       | Stroke of border around button                                                                   | 1                                                                                                                                                                                 |
| borderColor                       | color of border                                                                   | color of button                                                                                                                                                                                 |
| boxShadow                       | shadow of button                                                                   | This is an optional parameter                                                                                                                                                                                   |
| buttonColor                       | color of button                                                                   | Colors.blue                                                                                                                                                                                  |
| onPressed                       | Void call-back of button                                                                  | override this for listening the call back when button is pressed                                                                                                                                                                                  |
| tilt                       | controls the parallelogram tilt with max = 70 or min = -70                                                                   | 10                                                                                                                                                                                 |

[more detail](https://github.com/ibtihaajKhurram/shaped_buttons/tree/master/example/lib)


CHANGELOG.md
## [0.0.2] - Released on Feb 01, 2021.

* Added Tilt parameter in parallelogram button.

## [0.0.1] - Released on Jan 26, 2021.

* Initial release.


lib/src/my_parallelogram.dart
import 'package:flutter/material.dart';

class MyParallelogram extends CustomPainter {
  final double borderWidth;
  final Color buttonColor;
  final Color borderColor;
  final double tilt;

  MyParallelogram(
      {this.borderWidth = 1,
      this.buttonColor = Colors.white,
      this.tilt = 10,
      this.borderColor = Colors.white});
  @override
  void paint(Canvas canvas, Size size) {
    assert(this.tilt > -71 && this.tilt < 71,
        "The tilt value can't be more than 70 or less than -70");
    Paint paint_0 = new Paint()
      ..color = borderColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = borderWidth;

    Path path_0 = Path();
    path_0.moveTo(tilt, 0);
    path_0.lineTo(size.width, 0);
    path_0.lineTo(size.width - tilt, size.height);
    path_0.lineTo(0, size.height);
    path_0.close();

    canvas.drawPath(path_0, paint_0);

    Paint paint_1 = new Paint()
      ..color = buttonColor
      ..style = PaintingStyle.fill
      ..strokeWidth = 0;

    Path path_1 = Path();
    path_1.moveTo(tilt, 0);
    path_1.lineTo(size.width, 0);
    path_1.lineTo(size.width - tilt, size.height);
    path_1.lineTo(0, size.height);
    path_1.close();

    canvas.drawPath(path_1, paint_1);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}


lib/src/my_parallelogram_button.dart
import 'package:flutter/material.dart';
import 'package:shaped_buttons/src/my_parallelogram.dart';

class ParallalogramButton extends StatelessWidget {
  ParallalogramButton(
      {this.key,
      this.onPressed,
      this.buttonColor = Colors.blue,
      this.borderColor,
      this.borderWdth = 1,
      this.margin,
      this.height = 50,
      this.width = 120,
      this.boxShadow,
      this.tilt=10,
      @required this.child})
      : super(key: key);
  final Key key;
  final Widget child;
  final VoidCallback onPressed;
  final Color buttonColor;
  final List<BoxShadow> boxShadow;
  final Color borderColor;
  final double borderWdth;
  final EdgeInsetsGeometry margin;
  final double height;
  final double width;
  final double tilt;

  @override
  Widget build(BuildContext context) {
    return new Container(
      height: height,
      width: width,
      margin: margin ?? EdgeInsets.only(left: 1, right: 1),
      decoration: BoxDecoration(boxShadow: boxShadow),
      child: GestureDetector(
        onTap: onPressed,
        child: CustomPaint(
          painter: MyParallelogram(
              tilt: tilt,
              borderColor: borderColor ?? buttonColor,
              borderWidth: borderWdth,
              buttonColor: buttonColor),
          child: Center(
            child: Container(
              child: child,
            ),
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:shaped_buttons/shaped_buttons.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
        // This makes the visual density adapt to the platform that you run
        // the app on. For desktop platforms, the controls will be smaller and
        // closer together (more dense) than on mobile platforms.
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: DemoPage(),
    );
  }
}

class DemoPage extends StatelessWidget {
  const DemoPage({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Shaped Button usage demo'),
      ),
      body: Container(
          width: MediaQuery.of(context).size.width,
          child: Column(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              ParallalogramButton(
                buttonColor: Colors.yellow[700],
                child: Text(
                  "Regular Button",
                  style: TextStyle(color: Colors.black),
                ),
                onPressed: () {
                  print('Button Pressed');
                },
              ),
              ParallalogramButton(
                width: 200,
                borderColor: Colors.black,
                buttonColor: Colors.yellow[700],
                borderWdth: 5,
                child: Text(
                  "Button with border",
                  style: TextStyle(color: Colors.black),
                ),
                onPressed: () {
                  print('Button Pressed');
                },
              ),
              ParallalogramButton(
                width: 220,
                borderColor: Colors.black,
                buttonColor: Colors.transparent,
                borderWdth: 1,
                child: Text(
                  "Transparent button with border",
                  style: TextStyle(color: Colors.black),
                ),
                onPressed: () {
                  print('Button Pressed');
                },
              ),
              ParallalogramButton(
                width: 260,
                borderColor: Colors.black,
                buttonColor: Colors.yellow[700],
                tilt: 20,
                boxShadow: [
                  new BoxShadow(
                    blurRadius: 34.0,
                    color: Color.fromRGBO(0, 0, 0, 0.30),
                    offset: new Offset(0.0, 12.0),
                  ),
                ],
                borderWdth: 2,
                child: Text(
                  "Button with shadow and custom tilt",
                  style: TextStyle(color: Colors.black),
                ),
                onPressed: () {
                  print('Button Pressed');
                },
              ),
            ],
          )),
    );
  }
}
